### DISCUSSION TITLE
Feature Request: Implicit `virtual_table_name` when chaining in the Relational API

### DISCUSSION THREAD

**Discussion by ajfriend** [duckdb/duckdb#13672] [Ideas]
Hi folks, I'm really enjoying playing around with [DuckDB Relational API in Python](https://duckdb.org/docs/api/python/relational_api) as a way to incrementally build up queries.

However, the `DuckDBPyRelation.query()` function requires a `virtual_table_name` string that can be a little cumbersome to work with, and I'm wondering if it would be easy to add functionality to the API to make it unnecessary in some situations.

For example, I can load the iris data set and chain several SQL statements together with something like the following:

```python
import duckdb
iris = duckdb.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')

(iris
 .query('t1', 'from t1 select species, sepal_length as sl, sepal_width as sw')
 .query('t2', 'from t2 where sl > 4.5')
 .query('t3', 'from t3 where sw between 3.1 and 3.5')
 .query('t4', 'from t4 select species, avg(sl), max(sw) group by all')
 .query('t5', 'from t5 order by 2')
 .query('t6', 'from t6 limit 2')
)
```

However, each statement requires the somewhat repetitive pattern of `.query(<unique_table_name>, 'from <unique_table_name> ...')`. When I've done this in the past, I've found it error-prone and tricker to reorder or refactor statements.

Ideally, I'm thinking we could drop the `from <unique_table_name>` part, making it implied and unique so that we avoid name conflicts. For example, we could do some monkey patching to add a `DuckDBPyRelation.sql()` function (I'm surprised this name is still free!) like the following:

```python
from duckdb.duckdb import DuckDBPyRelation

def _sql(self: DuckDBPyRelation, stmt: str) -> DuckDBPyRelation:
    """
    Run the given SQL statement on the relation object, with an implied `FROM <temp_table_name>` clause.
    """
    # create a temporary table name
    # TODO: ideally, duckdb would do this for us, so we can avoid the need to come up with a unique name
    import random
    import string
    temp_table_name = '_tbl_' + ''.join(random.choices(string.ascii_lowercase, k=10))

    # insert the implied `from <temp_table_name>`, followed by the given SQL statement
    rel = self.query(
        temp_table_name,
        f'from {temp_table_name}\n{stmt}'
    )
    
    return rel

DuckDBPyRelation.sql = _sql
```

The chaining example above then simplifies to the following, without the need to come up with and specify table names:

```python
(iris
 .sql('select species, sepal_length as sl, sepal_width as sw')
 .sql('where sl > 4.5')
 .sql('where sw between 3.1 and 3.5')
 .sql('select species, avg(sl), max(sw) group by all')
 .sql('order by 2')
 .sql('limit 2')
)
```

While this works as a hack, it would be nice if functionality like this was included in DuckDB by default. I'm also assuming you could do something more intelligent to uniquely identify tables than my `temp_table_name = '_tbl_' + ''.join(random.choices(string.ascii_lowercase, k=10))`.

I've found the pattern above super ergonomic to work with, especially when I'm doing exploratory data analysis and can just write short SQL snippets like `where sl > 4.5`, `order by 2`, or `limit 2`. This functionality would also be helpful to someone who wanted to do [something funky with Python magic methods](https://gist.github.com/ajfriend/eea0795546c7c44f1c24ab0560a846b9) :)

Would it make sense to add a function along these lines to the relational API?

**Comment by Tishj** [comment]
While this is not exactly what you're looking for, repeating the table names can be avoided by using replacement scans:

```py
iris = duckdb.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/iris.csv')

iris = duckdb.sql("from iris select species, sepal_length as sl, sepal_width as sw")
iris = duckdb.sql("from iris where sl > 4.5")
...

**Comment by Tishj** [comment]
Using `iris = iris.select("species, sepal_length as sl, sepal_width as sw")` should also work I believe

**Comment by ajfriend** [comment]
Thanks! I had looked at those options as well, and they're good to know about, but they still don't quite achieve what I was hoping for. 

- I'd still like to be able to avoid typing `from <table_name>` if it can be inferred from the table.
- And the `iris = duckdb.sql("from iris where sl > 4.5")` example isn't ideal because it overwrites `iris`, which I may want to use unaltered later on. 
- Your `iris = iris.select()` example makes sense, but that only works for select statements; I'd love a solution that works for general SQL snippets.
- Also I usually set `config = {'enable_external_access': False}` to disable replacement scans so that my code is more explicit about which tables are being used, so it would be nice to have an option that doesn't depend on replacement scans.

Like I said, I don't think this is too hard for a user to cook up a way to do this in their own code. It just seems that a method like the `.sql()` above would be so generally useful in allowing users to execute concise or arbitrarily complex SQL with a single method that I wanted to float it by you all to see if you'd consider adding it to DuckDB.

Would it be much work to add? Or are there reasons not to add such a method?

**Reply by ajfriend** [reply]
Another way I could pitch it would be as an extension of [friendly SQL](https://duckdb.org/2022/05/04/friendlier-sql.html). The proposal above is, I think, pretty close to what was being proposed in [Google's pipe syntax SQL paper](https://research.google/pubs/sql-has-problems-we-can-fix-them-pipe-syntax-in-sql/) (but we can do it in DuckDB with a pretty small change!), which seems like it has a lot of benefits to improving the usability of SQL.

Of course, you can already do this with some work, but it isn't an obvious way to organize code with the current API. A change like this would nudge/hint users towards this "sequence of composable SQL snippets" approach. And, with DuckDB relations, I'm hoping we can optimize that sequence into an efficient execution plan.

**Reply by leostimpfle** [reply]
Just bumping this because I agree that this would be an incredible useful feature to have to build data pipelines in Python.

**Reply by NickCrews** [reply]
Let's move the discussion to the duckdb-python repo: https://github.com/duckdb/duckdb-python/discussions/357
